Downsample: C implementation for average, sum

gvtulder opened the pull request:
Hi. I'd like to propose the following changes to the downsampling code:
* extend the C implementation of Downsample to support `average_inc_pad` and `average_exc_pad`,
* add a new `sum` mode to the Python and C implementations.

The C implementation for the `average_*` modes reuses most of the code from `max`. The `sum` mode is a relatively straightforward extension.

nouiz commented:
Thanks for the PR. I'm starting to review it. travis have this error, can you look at it?<
```
======================================================================
FAIL: test_DownsampleFactorMax (theano.tensor.signal.tests.test_downsample.TestDownsampleFactorMax)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tensor/signal/tests/test_downsample.py", line 192, in test_DownsampleFactorMax
    assert numpy.all(output_val == numpy_output_val)
AssertionError
```

nouiz commented:
I finished my review, thanks for the PR. Mostly, you just need to fix the failing test on travis and rebase to make sure that this PR get tested with the c code for the grad with padding and sum.

gvtulder commented:
Thanks for the comments. I changed the test to use `assert_allclose`, which resolved the error. Perhaps the order of the summation is slightly different between the two numpy implementations.

I've rebased the branch to include the c code for the grad. There were no conflicts, probably because I did not implement the c version of grad for sum.

Since some of the tests for DownsampleFactorMaxGrad did not check average or sum, I updated the tests to test these extra modes.
